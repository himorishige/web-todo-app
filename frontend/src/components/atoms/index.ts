export { default as Button } from './Button';
export { default as StarIcon } from './StarIcon';
export { default as Input } from './Input';
export { default as CheckBox } from './CheckBox';
export { default as TextArea } from './TextArea';
export { default as TrashIcon } from './TrashIcon';
export { default as Spinner } from './Spinner';
export { default as HomeIcon } from './HomeIcon';
