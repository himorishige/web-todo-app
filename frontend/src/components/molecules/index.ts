export { default as TaskItem } from './TaskItem';
export { default as Message } from './Message';
