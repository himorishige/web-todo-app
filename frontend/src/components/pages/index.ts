export { default as Home } from './Home';
export { default as Page404 } from './Page404';
