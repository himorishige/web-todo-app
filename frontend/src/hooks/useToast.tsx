/** @jsxImportSource @emotion/react */
import React from 'react';
import { css, keyframes } from '@emotion/react';

type TOAST_TYPE = 'SUCCESS' | 'FAIL';

const ToastContext = React.createContext({
  showToast: (toastType: TOAST_TYPE, message: string) => {
    console.log(message);
  },
});

type Props = {
  children: React.ReactNode;
};

export const ToastProvider: React.VFC<Props> = (props) => {
  const [state, update] = React.useState({
    show: false,
    message: 'message',
    toastType: 'SUCCESS',
  });

  const showToast = (toastType: TOAST_TYPE, message: string) => {
    update({ show: true, message, toastType });
    setTimeout(() => update({ ...state, show: false, message: message }), 4000);
  };

  return (
    <>
      <ToastContext.Provider value={{ showToast }}>{props.children}</ToastContext.Provider>
      {state.show && (
        <div css={[toastStyle, state.toastType === 'FAIL' && failStyle]}>{state.message}</div>
      )}
    </>
  );
};

export const useToast = () => React.useContext(ToastContext);

const FadeIn = keyframes`
from {
  opacity: 0;
  top: 1rem;
}
to {
  opacity: 1;
  top: 2rem;
}
`;

const FadeOut = keyframes`
from {
  opacity: 1;
  top: 2rem;
}
to {
  opacity: 0;
  top: 1rem;
}
`;

const toastStyle = css`
  animation: ${FadeIn} 0.3s, ${FadeOut} 0.3s 3s forwards;
  position: fixed;
  display: flex;
  align-items: center;
  justify-content: center;
  min-width: 200px;
  min-height: 40px;
  background: var(--primary-color);
  padding: 1rem;
  top: 2rem;
  left: 50%;
  transform: translate(-50%, 0);
  border-radius: 4px;
  color: white;
`;

const failStyle = css`
  background: red;
`;
