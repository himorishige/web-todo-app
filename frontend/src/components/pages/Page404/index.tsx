/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { Layout } from 'src/components/templates';

type Props = {};

const Page404: React.VFC<Props> = () => {
  return (
    <Layout>
      <div css={footerStyle}>404</div>
    </Layout>
  );
};

export default Page404;

const footerStyle = css`
  display: flex;
  align-items: center;
  padding: 1rem;
`;
