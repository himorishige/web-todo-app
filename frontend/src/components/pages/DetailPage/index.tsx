/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { RouteComponentProps, useHistory } from 'react-router-dom';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { useForm, SubmitHandler } from 'react-hook-form';
import { Button, TextArea, TrashIcon } from 'src/components/atoms';
import { Message, TaskItem } from 'src/components/molecules';
import { Layout } from 'src/components/templates';
import {
  fetchAllTasks,
  removeTask,
  selectStatus,
  selectTasks,
  updateTask,
} from 'src/features/tasks/tasksSlice';
import { useToast } from 'src/hooks/useToast';
import { format } from 'date-fns';
import { useCallback, useEffect } from 'react';
import { memo } from 'react';

type Props = RouteComponentProps & {
  match: {
    params: {
      id: string;
    };
  };
};

type Inputs = {
  taskMemo: string;
};

const DetailPage: React.VFC<Props> = memo((props) => {
  const taskId = props.match.params.id;
  const status = useAppSelector(selectStatus);
  const task = useAppSelector((state) => selectTasks.selectById(state, taskId));
  const dispatch = useAppDispatch();
  const { showToast } = useToast();
  const history = useHistory();

  const { register, handleSubmit, watch } = useForm<Inputs>();

  // 完了フラグの管理
  const completedStateHandler = useCallback(
    async (id: string, isCompleted: boolean) => {
      const result = await dispatch(
        updateTask({
          id: id,
          isCompleted: isCompleted,
        }),
      );

      if (updateTask.rejected.match(result)) {
        showToast('FAIL', '完了フラグの更新に失敗しました');
      }
    },
    [dispatch, showToast],
  );

  // 優先度フラグの管理
  const priorityStateHandler = useCallback(
    async (id: string, priority: number) => {
      const result = await dispatch(
        updateTask({
          id: id,
          priority: priority,
        }),
      );

      if (updateTask.rejected.match(result)) {
        showToast('FAIL', '優先度の更新に失敗しました');
      }
    },
    [dispatch, showToast],
  );

  const onSubmit: SubmitHandler<Inputs> = useCallback(
    async (data) => {
      const result = await dispatch(
        updateTask({
          id: taskId,
          description: data.taskMemo,
        }),
      );

      if (updateTask.fulfilled.match(result)) {
        showToast('SUCCESS', 'メモを更新しました');
      }

      if (updateTask.rejected.match(result)) {
        showToast('FAIL', 'メモの更新に失敗しました');
      }
    },
    [dispatch, showToast, taskId],
  );

  const confirmPromise = useCallback((message: string) => {
    if (window.confirm(message)) {
      return Promise.resolve();
    }
    return Promise.reject();
  }, []);

  const deleteHandler = useCallback(async () => {
    confirmPromise('タスクを削除してもよいですか？').then(async () => {
      await dispatch(removeTask(taskId));

      showToast('SUCCESS', 'タスクを削除しました');
      history.push('/');
    });
  }, [confirmPromise, dispatch, history, showToast, taskId]);

  useEffect(() => {
    if (!task) {
      dispatch(fetchAllTasks());
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  return (
    <Layout>
      <div css={wrapperStyle}>
        {task ? (
          <form onSubmit={handleSubmit(onSubmit)}>
            <TaskItem
              id={task.id}
              title={task.title}
              description={task.description}
              isCompleted={task.isCompleted}
              priority={task.priority}
              completedStateHandler={completedStateHandler}
              priorityStateHandler={priorityStateHandler}
            />
            <div css={textAreaWrapper}>
              <TextArea
                label="taskMemo"
                placeholder="メモ"
                register={register}
                defaultValue={task.description}
              />
            </div>
            <div css={timeWrapper}>
              <div>登録：{format(new Date(task.createdAt), 'yyyy/MM/dd HH:mm')}</div>
              <div>更新：{format(new Date(task.updatedAt), 'yyyy/MM/dd HH:mm')}</div>
            </div>
            <div css={controlWrapper}>
              <div onClick={deleteHandler}>
                <TrashIcon />
              </div>
              <div>
                <Button primary label="メモを更新" disabled={status === 'loading'} />
              </div>
            </div>
          </form>
        ) : (
          <div css={message} data-testid="tasks-error">
            <Message>該当するIDのタスクはありません</Message>
          </div>
        )}
      </div>
    </Layout>
  );
});

export default DetailPage;

const wrapperStyle = css`
  padding: 1.25rem;
`;

const textAreaWrapper = css`
  margin-top: 1.25rem;
`;

const timeWrapper = css`
  display: flex;
  justify-content: flex-end;
  div {
    font-size: 0.75rem;
    margin-left: 1.25rem;
  }
`;

const controlWrapper = css`
  width: 100%;
  margin-top: 1.25rem;
  padding-left: 0.5rem;
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const message = css``;
