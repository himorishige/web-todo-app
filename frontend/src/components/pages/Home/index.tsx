/** @jsxImportSource @emotion/react */
import React, { useEffect } from 'react';
import { css } from '@emotion/react';
import { useAppDispatch, useAppSelector } from 'src/app/hooks';
import { Spinner } from 'src/components/atoms';
import { TasksList } from 'src/components/organisms';
import { Layout } from 'src/components/templates';
import { fetchAllTasks, selectTasks } from 'src/features/tasks/tasksSlice';

const Home: React.VFC = () => {
  const dispatch = useAppDispatch();
  const tasks = useAppSelector(selectTasks.selectAll);

  useEffect(() => {
    const init = async () => {
      const result = await dispatch(fetchAllTasks());

      if (fetchAllTasks.fulfilled.match(result)) {
        console.log('useEffect Updated');
      }

      if (fetchAllTasks.rejected.match(result)) {
        console.log('useEffect not updated');
      }
    };
    if (!tasks.length) {
      init();
    }
  }, [dispatch, tasks]);

  return (
    <Layout>
      <div css={wrapperStyle}>
        {tasks.length ? (
          <TasksList />
        ) : (
          <div css={loadingStyle}>
            <Spinner />
          </div>
        )}
      </div>
    </Layout>
  );
};

export default Home;

const wrapperStyle = css`
  padding: 1.25rem;
`;

const loadingStyle = css`
  width: 100%;
  height: calc(100vh - 96px - 118px);
  display: flex;
  align-items: center;
  justify-content: center;
`;
