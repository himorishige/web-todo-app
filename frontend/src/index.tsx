import React from 'react';
import ReactDOM from 'react-dom';
import App from 'src/App';
import { store } from 'src/app/store';
import { Provider } from 'react-redux';
import { Global } from '@emotion/react';
import * as serviceWorker from 'src/serviceWorker';
import 'sanitize.css';
import { globalStyle } from 'src/styles/globalStyle';
import { ToastProvider } from './hooks/useToast';

ReactDOM.render(
  <React.StrictMode>
    <Provider store={store}>
      <Global styles={globalStyle} />
      <ToastProvider>
        <App />
      </ToastProvider>
    </Provider>
  </React.StrictMode>,
  document.getElementById('root'),
);

// If you want your app to work offline and load faster, you can change
// unregister() to register() below. Note this comes with some pitfalls.
// Learn more about service workers: https://bit.ly/CRA-PWA
serviceWorker.unregister();
