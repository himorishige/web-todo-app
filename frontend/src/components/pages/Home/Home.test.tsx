import { render, screen, cleanup } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { Provider } from 'react-redux';
import { configureStore } from '@reduxjs/toolkit';
import tasksReducer from '../../../features/tasks/tasksSlice';
import { HelmetProvider } from 'react-helmet-async';
import { Home } from 'src/components/pages';
import { ApiResponseType, Task } from 'src/types';
import { BrowserRouter } from 'react-router-dom';
import { API_URL } from 'src/constants';
import { ToastProvider } from 'src/hooks/useToast';
import userEvent from '@testing-library/user-event';

const mockData: ApiResponseType<Task[]> = {
  status: 'ok',
  data: [
    {
      createdAt: '2021-07-10T08:52:01.696Z',
      priority: 0,
      id: 'bbeff447-6b94-402d-8961-7ab44e9f6fc7',
      description: '',
      updatedAt: '2021-07-10T08:52:01.696Z',
      title: '卵を買ってくる',
      isCompleted: false,
    },
    {
      updatedAt: '2021-07-10T07:31:51.095Z',
      createdAt: '2021-07-09T12:09:03.565Z',
      priority: 1,
      description: 'メグミルク',
      id: '5672a87d-5129-4987-a579-fde08a8c5d41',
      isCompleted: false,
      title: '牛乳を買ってくる',
    },
  ],
};

const server = setupServer(
  rest.get(API_URL, (req, res, ctx) => {
    return res(ctx.status(200), ctx.json(mockData));
  }),
);

beforeAll(() => server.listen());
afterEach(() => {
  server.resetHandlers();
  cleanup();
});
afterAll(() => server.close());

describe('HomePage', () => {
  let store: any;
  beforeEach(() => {
    store = configureStore({
      reducer: {
        tasks: tasksReducer,
      },
    });
  });
  test('正しくレンダリングされている', () => {
    const target = render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(target).toMatchSnapshot();
  });
  test('APIから出力されたタスクが2件表示される', async () => {
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findAllByTestId('task-item')).toHaveLength(2);
  });

  test('タスクは上から作成日順に並んでいる', async () => {
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findAllByTestId('task-item')).toHaveLength(2);
    const target = await screen.findAllByTestId('task-item');
    expect(target[0]).toHaveTextContent('牛乳を買ってくる');
  });

  test('お気に入りタスクにフィルターした時に正しくフィルターされたタスクが表示される', async () => {
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findAllByTestId('task-item')).toHaveLength(2);
    userEvent.click(await screen.findByTestId('toggle-button'));
    expect(await screen.findAllByTestId('task-item')).toHaveLength(1);
  });

  test('完了フラグをつけた時に順番が入れ替わる', async () => {
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findAllByTestId('task-item')).toHaveLength(2);
    const target = await screen.findAllByTestId('task-item');
    expect(target[0]).toHaveTextContent('牛乳を買ってくる');
    const check = await screen.findAllByTestId('checkbutton');
    userEvent.click(check[0]);
    const nextTarget = await screen.findAllByTestId('task-item');
    expect(nextTarget[0]).toHaveTextContent('牛乳を買ってくる');
  });

  test('ネットワークエラーの時にエラーメッセージが表示される', async () => {
    server.use(
      rest.get(API_URL, (req, res, ctx) => {
        return res(ctx.status(404));
      }),
    );
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <Home />
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findByTestId('tasks-net-error')).toBeTruthy();
  });
  test('API拒否時にエラーが表示される', async () => {
    server.use(
      rest.get(API_URL, (req, res, ctx) => {
        return res(
          ctx.status(403),
          ctx.json({
            error: 'error',
          }),
        );
      }),
    );
    render(
      <Provider store={store}>
        <HelmetProvider>
          <BrowserRouter>
            <ToastProvider>
              <Home />
            </ToastProvider>
          </BrowserRouter>
        </HelmetProvider>
      </Provider>,
    );
    expect(screen.queryAllByTestId('tasks-item')).toHaveLength(0);
    expect(await screen.findByTestId('tasks-net-error')).toBeTruthy();
  });
});
