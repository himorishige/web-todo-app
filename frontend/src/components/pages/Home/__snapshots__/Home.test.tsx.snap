// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomePage 正しくレンダリングされている 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <div
          class="css-1hrd1c8"
        >
          <div
            class="css-1hh8z22"
          >
            <div>
              <a
                href="/"
              >
                <svg
                  class="css-1443oa1"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M14 1H2a1 1 0 00-1 1v12a1 1 0 001 1h12a1 1 0 001-1V2a1 1 0 00-1-1zM2 0a2 2 0 00-2 2v12a2 2 0 002 2h12a2 2 0 002-2V2a2 2 0 00-2-2H2z"
                    fill-rule="evenodd"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M15 4H1V3h14v1z"
                    fill-rule="evenodd"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M5 15.5v-14h1v14H5zm5 0v-14h1v14h-1z"
                    fill-rule="evenodd"
                  />
                  <path
                    clip-rule="evenodd"
                    d="M15 8H1V7h14v1zm0 4H1v-1h14v1z"
                    fill-rule="evenodd"
                  />
                  <path
                    d="M0 2a2 2 0 012-2h12a2 2 0 012 2v2H0V2z"
                  />
                </svg>
              </a>
            </div>
            <h1
              class="css-1llq2fx"
            >
              ToDo App
            </h1>
            <div
              class="css-knz8lr"
              data-testid="toggle-button"
            >
              <div
                class="css-kknodv"
              >
                <svg
                  class="css-ba990d"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    clip-rule="evenodd"
                    d="M2.866 14.85c-.078.444.36.791.746.593l4.39-2.256 4.389 2.256c.386.198.824-.149.746-.592l-.83-4.73 3.523-3.356c.329-.314.158-.888-.283-.95l-4.898-.696L8.465.792a.513.513 0 00-.927 0L5.354 5.12l-4.898.696c-.441.062-.612.636-.283.95l3.523 3.356-.83 4.73zm4.905-2.767l-3.686 1.894.694-3.957a.565.565 0 00-.163-.505L1.71 6.745l4.052-.576a.525.525 0 00.393-.288l1.847-3.658 1.846 3.658a.525.525 0 00.393.288l4.052.575-2.906 2.77a.564.564 0 00-.163.506l.694 3.957-3.686-1.894a.503.503 0 00-.461 0z"
                    fill-rule="evenodd"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
        <main
          class="css-1yvttm9-Layout"
        >
          <div
            class="css-4pg0u1-Home"
          >
            <div
              class="css-e7d0e4-Home"
            >
              <div>
                <div
                  aria-busy="true"
                  class=""
                >
                  <svg
                    aria-label="audio-loading"
                    height="100"
                    viewBox="0 0 38 38"
                    width="100"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <defs>
                      <lineargradient
                        id="a"
                        x1="8.042%"
                        x2="65.682%"
                        y1="0%"
                        y2="23.865%"
                      >
                        <stop
                          offset="0%"
                          stop-color="#5e454b"
                          stop-opacity="0"
                        />
                        <stop
                          offset="63.146%"
                          stop-color="#5e454b"
                          stop-opacity=".631"
                        />
                        <stop
                          offset="100%"
                          stop-color="#5e454b"
                        />
                      </lineargradient>
                    </defs>
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <g
                        transform="translate(1 1)"
                      >
                        <path
                          d="M36 18c0-9.94-8.06-18-18-18"
                          id="Oval-2"
                          stroke="#5e454b"
                          stroke-width="2"
                        >
                          <animatetransform
                            attributeName="transform"
                            dur="0.9s"
                            from="0 18 18"
                            repeatCount="indefinite"
                            to="360 18 18"
                            type="rotate"
                          />
                        </path>
                        <circle
                          cx="36"
                          cy="18"
                          fill="#fff"
                          r="1"
                        >
                          <animatetransform
                            attributeName="transform"
                            dur="0.9s"
                            from="0 18 18"
                            repeatCount="indefinite"
                            to="360 18 18"
                            type="rotate"
                          />
                        </circle>
                      </g>
                    </g>
                  </svg>
                </div>
              </div>
            </div>
          </div>
        </main>
        <footer
          class="css-t18ryb-Layout"
        >
          <form>
            <div
              class="css-1qqqf83"
            >
              <div
                class="css-7d0ne1"
              >
                <div
                  class="css-1ago99h"
                >
                  <input
                    class="css-5x3p9d"
                    disabled=""
                    maxlength="40"
                    name="taskName"
                    placeholder="新しいタスクを入力してください"
                    type="text"
                    value=""
                  />
                </div>
                <div
                  class="css-75wbz6"
                >
                  <button
                    class="css-k3l7h4"
                    disabled=""
                  >
                    登録
                  </button>
                </div>
              </div>
            </div>
          </form>
        </footer>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <div
        class="css-1hrd1c8"
      >
        <div
          class="css-1hh8z22"
        >
          <div>
            <a
              href="/"
            >
              <svg
                class="css-1443oa1"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M14 1H2a1 1 0 00-1 1v12a1 1 0 001 1h12a1 1 0 001-1V2a1 1 0 00-1-1zM2 0a2 2 0 00-2 2v12a2 2 0 002 2h12a2 2 0 002-2V2a2 2 0 00-2-2H2z"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M15 4H1V3h14v1z"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M5 15.5v-14h1v14H5zm5 0v-14h1v14h-1z"
                  fill-rule="evenodd"
                />
                <path
                  clip-rule="evenodd"
                  d="M15 8H1V7h14v1zm0 4H1v-1h14v1z"
                  fill-rule="evenodd"
                />
                <path
                  d="M0 2a2 2 0 012-2h12a2 2 0 012 2v2H0V2z"
                />
              </svg>
            </a>
          </div>
          <h1
            class="css-1llq2fx"
          >
            ToDo App
          </h1>
          <div
            class="css-knz8lr"
            data-testid="toggle-button"
          >
            <div
              class="css-kknodv"
            >
              <svg
                class="css-ba990d"
                fill="currentColor"
                height="1em"
                stroke="currentColor"
                stroke-width="0"
                viewBox="0 0 16 16"
                width="1em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  clip-rule="evenodd"
                  d="M2.866 14.85c-.078.444.36.791.746.593l4.39-2.256 4.389 2.256c.386.198.824-.149.746-.592l-.83-4.73 3.523-3.356c.329-.314.158-.888-.283-.95l-4.898-.696L8.465.792a.513.513 0 00-.927 0L5.354 5.12l-4.898.696c-.441.062-.612.636-.283.95l3.523 3.356-.83 4.73zm4.905-2.767l-3.686 1.894.694-3.957a.565.565 0 00-.163-.505L1.71 6.745l4.052-.576a.525.525 0 00.393-.288l1.847-3.658 1.846 3.658a.525.525 0 00.393.288l4.052.575-2.906 2.77a.564.564 0 00-.163.506l.694 3.957-3.686-1.894a.503.503 0 00-.461 0z"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </div>
        </div>
      </div>
      <main
        class="css-1yvttm9-Layout"
      >
        <div
          class="css-4pg0u1-Home"
        >
          <div
            class="css-e7d0e4-Home"
          >
            <div>
              <div
                aria-busy="true"
                class=""
              >
                <svg
                  aria-label="audio-loading"
                  height="100"
                  viewBox="0 0 38 38"
                  width="100"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <defs>
                    <lineargradient
                      id="a"
                      x1="8.042%"
                      x2="65.682%"
                      y1="0%"
                      y2="23.865%"
                    >
                      <stop
                        offset="0%"
                        stop-color="#5e454b"
                        stop-opacity="0"
                      />
                      <stop
                        offset="63.146%"
                        stop-color="#5e454b"
                        stop-opacity=".631"
                      />
                      <stop
                        offset="100%"
                        stop-color="#5e454b"
                      />
                    </lineargradient>
                  </defs>
                  <g
                    fill="none"
                    fill-rule="evenodd"
                  >
                    <g
                      transform="translate(1 1)"
                    >
                      <path
                        d="M36 18c0-9.94-8.06-18-18-18"
                        id="Oval-2"
                        stroke="#5e454b"
                        stroke-width="2"
                      >
                        <animatetransform
                          attributeName="transform"
                          dur="0.9s"
                          from="0 18 18"
                          repeatCount="indefinite"
                          to="360 18 18"
                          type="rotate"
                        />
                      </path>
                      <circle
                        cx="36"
                        cy="18"
                        fill="#fff"
                        r="1"
                      >
                        <animatetransform
                          attributeName="transform"
                          dur="0.9s"
                          from="0 18 18"
                          repeatCount="indefinite"
                          to="360 18 18"
                          type="rotate"
                        />
                      </circle>
                    </g>
                  </g>
                </svg>
              </div>
            </div>
          </div>
        </div>
      </main>
      <footer
        class="css-t18ryb-Layout"
      >
        <form>
          <div
            class="css-1qqqf83"
          >
            <div
              class="css-7d0ne1"
            >
              <div
                class="css-1ago99h"
              >
                <input
                  class="css-5x3p9d"
                  disabled=""
                  maxlength="40"
                  name="taskName"
                  placeholder="新しいタスクを入力してください"
                  type="text"
                  value=""
                />
              </div>
              <div
                class="css-75wbz6"
              >
                <button
                  class="css-k3l7h4"
                  disabled=""
                >
                  登録
                </button>
              </div>
            </div>
          </div>
        </form>
      </footer>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
